File: web/source/hackernews/submissions.js
// JS port of sources/hackernews/submissions.py.
//
// Walks the Algolia HN Search API for stories submitted by a
// username, newest-first, with the same page-level early exit.
//
// Upvotes (sources/hackernews/upvotes.py) is deliberately omitted —
// it needs the HN password that lives encrypted server-side.

import { fetchJson, sleep } from "../utils/http.js";
import { isoDate } from "../utils/doc.js";

const ALGOLIA = "https://hn.algolia.com/api/v1/search_by_date";
const DELAY_MS = 500;

export async function submissions({
  username,
  maxItems = 500,
  existingUrls = null,
}) {
  const out = {};
  let page = 0;
  let total = 0;
  while (total < maxItems) {
    const url = `${ALGOLIA}?tags=story,author_${encodeURIComponent(
      username,
    )}&hitsPerPage=100&page=${page}`;
    let result;
    try {
      result = await fetchJson(url);
    } catch (err) {
      console.warn("[sync] HN submissions algolia error:", err);
      break;
    }
    const hits = result.hits || [];
    if (hits.length === 0) break;

    if (existingUrls) {
      const newInPage = hits.filter(
        (h) => h.url && !existingUrls.has(h.url),
      ).length;
      if (newInPage === 0) break;
    }

    for (const hit of hits) {
      const storyUrl = hit.url;
      if (!storyUrl) continue;
      if (existingUrls && existingUrls.has(storyUrl)) continue;
      if (out[storyUrl]) continue;
      const title = hit.title || "";
      const created = isoDate(hit.created_at);
      const points = hit.points || 0;
      const nComments = hit.num_comments || 0;
      const summary =
        points || nComments
          ? `${points} points, ${nComments} comments on HN`
          : "";
      out[storyUrl] = {
        title,
        summary,
        date: created,
        tags: ["hackernews"],
        source: "hackernews",
      };
    }

    total += hits.length;
    page += 1;
    if (page >= (result.nbPages || 0)) break;
    await sleep(DELAY_MS);
  }
  return out;
}

File: sources/hackernews/submissions.py
"""
HackerNews submissions extractor.

Fetches stories a user submitted to HN. These are often their own
blog posts or projects they find important enough to share.
Uses the public Algolia API — no auth needed.
"""

import json
import time
import urllib.request

__all__ = ["Submissions"]

_ALGOLIA = "https://hn.algolia.com/api/v1/search_by_date"
_DELAY = 0.5


def _fetch_json(url: str) -> dict:
    req = urllib.request.Request(url, headers={"User-Agent": "Knowledge/1.0"})
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read())


class Submissions:
    """Fetch stories submitted by a HN user."""

    def __init__(self, username: str, max_items: int = 500):
        self.username = username
        self.max_items = max_items

    def __call__(self, existing_urls: set[str] | None = None) -> dict[str, dict]:
        print(f"    Fetching HN submissions for @{self.username}...")
        data: dict[str, dict] = {}
        page = 0
        total = 0

        while total < self.max_items:
            url = f"{_ALGOLIA}?tags=story,author_{self.username}&hitsPerPage=100&page={page}"
            try:
                result = _fetch_json(url)
            except Exception as e:
                print(f"    Algolia error: {e}")
                break

            hits = result.get("hits", [])
            if not hits:
                break

            # Page-level early exit: Algolia returns most-recent-first. If every
            # story on this page is already known and has a URL, we are caught up.
            if existing_urls is not None:
                new_in_page = sum(1 for h in hits if h.get("url") and h["url"] not in existing_urls)
                if new_in_page == 0:
                    print(f"    No new submissions on page {page}, stopping early.")
                    break
                print(f"    Page {page}: {new_in_page} new submission(s).")

            for hit in hits:
                story_url = hit.get("url")
                if not story_url:
                    continue
                if existing_urls and story_url in existing_urls:
                    continue
                if story_url in data:
                    continue
                title = hit.get("title") or ""
                created = (hit.get("created_at") or "")[:10]
                points = hit.get("points") or 0
                num_comments = hit.get("num_comments") or 0

                summary = ""
                if points or num_comments:
                    summary = f"{points} points, {num_comments} comments on HN"

                data[story_url] = {
                    "title": title,
                    "summary": summary,
                    "date": created,
                    "tags": ["hackernews"],
                }

            total += len(hits)
            page += 1
            if page >= result.get("nbPages", 0):
                break
            time.sleep(_DELAY)

        print(f"    {len(data)} stories submitted")
        return data
